iceoryx_utils/include/iceoryx_utils/internal/relocatable_pointer/relative_ptr.hpp

```diff
@@ -13,291 +14,95 @@
-    {
-        return getRepository().getBasePtr(id);
-    }
-
-    ///@brief unregister all ptr id pairs (leads to initial state)
-    static void unregisterAll()
-    {
-        getRepository().unregisterAll();
-    }
-
-    ///@brief get the offset from id and ptr
-    ///@return offset
-    static offset_t getOffset(const id_t id, const_ptr_t ptr)
-    {
-        if (id == NULL_POINTER_ID)
-        {
-            return NULL_POINTER_OFFSET;
-        }
-        auto basePtr = getBasePtr(id);
-        return reinterpret_cast<offset_t>(ptr) - reinterpret_cast<offset_t>(basePtr);
-    }
-
-
-    ///@brief get the pointer from id and offset ("inverse" to getOffset)
-    ///@return ptr
-    static ptr_t getPtr(const id_t id, const offset_t offset)
-    {
-        if (offset == NULL_POINTER_OFFSET)
-        {
-            return nullptr;
-        }
-        auto basePtr = getBasePtr(id);
-        return reinterpret_cast<ptr_t>(offset + reinterpret_cast<offset_t>(basePtr));
-    }
-
-    static id_t searchId(ptr_t ptr)
-    {
-        if (ptr == nullptr)
-        {
-            return NULL_POINTER_ID;
-        }
-        return getRepository().searchId(ptr);
-    }
-
-    static bool isValid(id_t id)
-    {
-        return getRepository().isValid(id);
-    }
-
-    static PointerRepository<id_t, ptr_t>& getRepository() noexcept
-    {
-        static PointerRepository<id_t, ptr_t> repository;
-        return repository;
-    }
-
-    //*****************************************************************************************
-
-    offset_t computeOffset(ptr_t ptr)
-    {
-        return getOffset(m_id, ptr);
-    }
-
-    ptr_t computeRawPtr() const
-    {
-        return getPtr(m_id, m_offset);
-    }
-
-    static constexpr id_t NULL_POINTER_ID = std::numeric_limits<id_t>::max();
-    static constexpr offset_t NULL_POINTER_OFFSET = std::numeric_limits<offset_t>::max();
-
-  protected:
-    id_t m_id{NULL_POINTER_ID};
-    offset_t m_offset{NULL_POINTER_OFFSET};
-};
-
+/// @brief typed version so we can use operator->
 template <typename T>
-class relative_ptr : public RelativePointer
+class RelativePointer : public BaseRelativePointer
 {
   public:
-    relative_ptr(ptr_t ptr, id_t id)
-        : RelativePointer(ptr, id)
-    {
-    }
-
-    relative_ptr(offset_t offset, id_t id)
-        : RelativePointer(offset, id)
-    {
-    }
-
-    relative_ptr(ptr_t ptr = nullptr)
-        : RelativePointer(ptr)
-    {
-    }
-
-
-    relative_ptr(const RelativePointer& other)
-        : RelativePointer(other)
-    {
-    }
-
-    relative_ptr& operator=(const RelativePointer& other)
-    {
-        RelativePointer::operator=(other);
-
-        return *this;
-    }
-
-    relative_ptr& operator=(ptr_t ptr)
-    {
-        m_id = searchId(ptr);
-        m_offset = computeOffset(ptr);
-
-        return *this;
-    }
-
+    /// @brief constructs a RelativePointer pointing to ptr in a segment identified by id
+    /// @param[in] ptr is the pointee
```

Comment:
I think I understand what you mean with this now: the RelativePointer points to ptr (it locates the ptr) which in turn points to the data. Therefore ptr is the pointee of a RelativePointer. It can certainly be seen like this.

On the other hand my view is: they logically both point to the same data (the pointee), and can be constrcuted from one another (here relativePointer from ptr). Neither are really pointees in my book if we see it this way.

What is more "natural" or intuitive I cannot say. We just need to be consistent. I prefer the second interpretation, it is more in line of how we think smart pointers like shared_ptr and so on.